All of Review - Ansu V.

aavarghese opened the pull request:
A fake PR for reviewing purposes...
<!--

Thanks for submitting a PR to OPA!

Before pressing 'Create pull request' please read the checklist below.

* All code changes should be accompanied with tests. If you are not
modifying any tests, just provide a short explanation of why updates
to tests are not necessary. In addition to helping catch bugs, tests
are extremely helpful in providing _context_ that explains how your
changes can be used.

* All changes to public APIs **must** be accompanied with
docs. Examples of public APIs include built-in functions,
config fields, and of course, exported Go types/functions/constants/etc.

* Commit messages should explain _why_ you made the changes, not what
you changed. Use active voice. Keep the subject line under 50
characters or so.

* All commits must be signed off by the author. If you are not
familiar with signing off, see CONTRIBUTING.md below.

For more information on contributing to OPA see:

* [CONTRIBUTING.md](https://github.com/open-policy-agent/opa/blob/main/CONTRIBUTING.md)
  for high-level contribution guidelines.

* [DEVELOPMENT.md](https://github.com/open-policy-agent/opa/blob/main/docs/devel/DEVELOPMENT.md)
  for development workflow and environment setup.

-->

juliafriedman8 committed:
ast: support for allOf in JSON schema type checker

Currently OPA's schema checker does not support allOf. This commit parses allOf by merging all the subschemas into a single schema before converting to a Rego type. I also included a short explanation in schemas.md.

Related to: open-policy-agent#3592

Signed-off-by: Julia Friedman <juliafriedman@ibm.com>

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfSchemaParentVariation,
+		},
+		{
+			note:   "empty schema with allOf",
```
Can you change this to "allOf with empty schema"

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfArraySchema,
+		},
+		{
+			note:   "allOf schema with parent variation",
```
Can you explain what the variation is?

juliafriedman8 replied:
is "allOf schema with no parents" good?

aavarghese replied:
`allOf without a parent schema` maybe?

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+		schema string
+	}{
+		{
+			note:   "allOf object schema",
```
Maybe "allOf with mergeable Object types in schema"?

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfObjectSchema,
+		},
+		{
+			note:   "allOf array schema",
```
Similar to above...

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: emptySchema,
+		},
+		{
+			note:   "allOf schema with array of objects",
```
"allOf with mergeable Array of Object types in schema"?

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfArrayOfObjects,
+		},
+		{
+			note:   "allOf object schema with type declaration missing",
```
"allOf with mergeable Object types in schema with type declaration missing"?

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfObjectMissing,
+		},
+		{
+			note:   "allOf array schema with different types within the array",
```
Similar to above.

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfArrayDifTypes,
+		},
+		{
+			note:   "allOf object schema with an array inside object",
```
Similar to above.

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfArrayInsideObject,
+		},
+		{
+			note:   "allOf array schema with type declaration missing",
```
Similar to above.

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfIntegerSchema,
+		},
+		{
+			note:   "allOf boolean schema",
```
"allOf with mergeable Boolean types in schema"

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfArrayMissing,
+		},
+		{
+			note:   "allOf inside of core schema",
```
Similar to above.

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfInsideCoreSchema,
+		},
+		{
+			note:   "allOf string schema",
```
Similar to above.

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			schema: allOfStringSchema,
+		},
+		{
+			note:   "allOf integer schema",
```
Similar to above.

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+	}
+}
+
+func TestCompilerCheckTypesWithAllOfSchemaErrors(t *testing.T) {
```
This doesn't have to be a separate test method. You can add `expectedError` to the struct in the test method above too, and the test cases without errors don't need `expectedError` set - thats all.

Inside the t.Run....you can separate the two implementations by adding an `if tc.expectedError != nil` condition

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,130 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+		{
+			note:          "allOf schema with array of arrays",
+			schema:        allOfArrayOfArrays,
+			expectedError: fmt.Errorf("unable to merge these schema"),
```
Can you modify all four error strings to be `unable to merge these schemas` <-- notice schemas instead of schema

aavarghese commented on ast/compile_test.go:
```diff
@@ -4348,56 +4472,6 @@ func TestWithSchema(t *testing.T) {
 	}
 }
 
-func TestAnyOfObjectSchema1(t *testing.T) {
```
All of Jing's test methods and some of her schemas at the bottom were removed accidentally during the merge. Please add them back

aavarghese replied:
Running `make test` fails because the merge accidentally deleted some of Jing's stuff

aavarghese commented on ast/compile_test.go:
```diff
@@ -4669,27 +4897,169 @@ const anyOfObjectMissing = `{
 	]
 }`
 
-const anyOfArraySchema = ` {
```
schema was deleted

juliafriedman8 replied:
this is actually there i added it back in when i saw it got deleted in the merge

aavarghese replied:
ok - maybe a bad git difference rendered by the UI. As long as `make test` passes, it should confirm all schemas are there. @juliafriedman8

aavarghese commented on ast/schema_test.go:
```diff
@@ -79,34 +84,235 @@ func TestSetTypesWithPodSchema(t *testing.T) {
+	objectSchemaStaticProps = append(objectSchemaStaticProps, &types.StaticProperty{Key: "ZipCode", Value: types.S})
+	objectSchemaStaticProps = append(objectSchemaStaticProps, &types.StaticProperty{Key: "County", Value: types.S})
+	objectSchemaStaticProps = append(objectSchemaStaticProps, &types.StaticProperty{Key: "PostCode", Value: types.S})
+	obejctSchemaExpectedType := types.NewObject(objectSchemaStaticProps, nil)
```
typo in obejctSchemaExpectedType

aavarghese commented on ast/schema_test.go:
```diff
@@ -79,34 +84,235 @@ func TestSetTypesWithPodSchema(t *testing.T) {
+	//Test 2: array schema
+	arrayExpectedType := types.NewArray(nil, types.N)
+
+	//Test 3: Parent variation
```
if all comments start with lowercase...can you make this `parent variation` to be consistent? Same for test cases below..

aavarghese commented on ast/schema_test.go:
```diff
@@ -79,34 +84,235 @@ func TestSetTypesWithPodSchema(t *testing.T) {
+	expectedStringType := types.NewString()
+	expectedIntegerType := types.NewNumber()
+	expectedBooleanType := types.NewBoolean()
+	expectedTypeError := fmt.Errorf("unable to merge these schemas")
```
you have this same error declared in four different variables above: objectAndArrayExpectedError, arrayDifTypesExpectedError etc....having just one var is good and re-using for all error cases.

aavarghese commented on docs/content/schemas.md:
```diff
@@ -529,6 +529,81 @@ Once this is fixed, the second typo is highlighted, prompting the user to choose
 
+#### `allOf`
+
+Specifically, `allOf` acts as an Rego And type where all of the subschemas must be true. Consider the following Rego and schema file containing `allOf`:
```
this will need to be re-worded :) "acts as an Rego And type" is wrong --> there is no Rego And type. Rego has an Or type which was used by anyOf...but thats not true for allOf which is why we had to implement the merge function.

aavarghese commented on docs/content/schemas.md:
````diff
@@ -529,6 +529,81 @@ Once this is fixed, the second typo is highlighted, prompting the user to choose
+	              have: "servers"
+	              want: ["server"]
+```
+Once this is fixed, the second typo is highlighted, informing the user that `versions` should be `version`. Also, the type checker will return an error in this case because `kind` and `version` are of type string, while `accessNum` is of type integer. Because all of the types must be true within `allOf`, this will generate an error due to different types.
````
Why is this generating an error? These are object properties -- they dont have to be the same type since you append them all together. I'm confused

juliafriedman8 replied:
oops in my head i think i was subconsciously treating this as an array

aavarghese commented on docs/content/schemas.md:
```diff
@@ -529,6 +529,81 @@ Once this is fixed, the second typo is highlighted, prompting the user to choose
+#### `allOf`
+
+Specifically, `allOf` acts as an Rego And type where all of the subschemas must be true. Consider the following Rego and schema file containing `allOf`:
+
```
You can add a one-line description (similar to comment in the Github issue) about merging the sub-schemas for allOf and converting the single schema to a Rego type.

aavarghese commented on ast/compile.go:
```diff
@@ -928,6 +982,42 @@ func parseSchema(schema interface{}) (types.Type, error) {
 		return orType, nil
 	}
 
+	//Handle allOf
```
Remove comment

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,60 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+	return schema.Types.String() == s
+}
+
+//Merges JSON schemas
```
remove comment - it is self explanatory

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,60 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
 	return schemasCompiled, nil
 }
 
+//Checks if a JSON schema contains a specific type
```
remove comment - it is self explanatory

aavarghese commented on ast/compile.go:
```diff
@@ -928,6 +982,42 @@ func parseSchema(schema interface{}) (types.Type, error) {
+		subSchemaArray := subSchema.AllOf
+		allOfResult, err := mergeSchemas(subSchemaArray)
+		if err != nil {
+			return nil, fmt.Errorf("unable to merge these schema due to: %w", err)
```
you can just do a `return nil, err` --> No need for another fmt.Errorf with unable to merge...since `err` already has those exact same words

aavarghese commented on ast/compile.go:
```diff
@@ -928,6 +982,42 @@ func parseSchema(schema interface{}) (types.Type, error) {
+					}
+					return parseSchema(arrayResult)
+				}
+			}
```
What if the parent type contains `integer`, `boolean`, `string` etc? Those types are not being handled.
For ex, give this schema a try
```
`{
	"$schema": "http://json-schema.org/draft-04/schema#",
	"type": "string",
	"title": "The b schema",
	"description": "An explanation about the purpose of this instance.",
	"allOf": [{
			"type": "integer",
		},
		{
			"type": "integer",
		}
	]
}`
```
It returns a new type of `integer` when it should have been failing with `unable to merge schemas` error

aavarghese commented on ast/compile.go:
```diff
@@ -928,6 +982,42 @@ func parseSchema(schema interface{}) (types.Type, error) {
+				var temp []*gojsonschema.SubSchema
+				temp = append(temp, allOfResult)
+				temp = append(temp, subSchema)
+				if len(subSchema.ItemsChildren) > 0 {
```
You dont need this if condition since mergeSchemas will only merge if there are children...

aavarghese commented on ast/compile.go:
```diff
@@ -928,6 +982,42 @@ func parseSchema(schema interface{}) (types.Type, error) {
+				return parseSchema(objectResult)
+			}
+
+			if subSchema.Types.Contains("array") && allOfResult.Types.Contains("array") {
```
This if condition doesnt need to be separate - it can be combined with the above if condition in one swoop
`if (subSchema.Types.Contains("object") && allOfResult.Types.Contains("object")) || (subSchema.Types.Contains("array") && allOfResult.Types.Contains("array"))`

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,60 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+
+//Merges JSON schemas
+func mergeSchemas(schemas []*gojsonschema.SubSchema) (*gojsonschema.SubSchema, error) {
+	var result = schemas[0]
```
This line needs to be after the len(schemas) == 0 check....otherwise schemas[0] will crash

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,60 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+			return nil, fmt.Errorf("unable to merge these schemas")
+		} else if result.Types.Contains("string") && schemas[i].Types.Contains("string") {
+			return result, nil
+		} else if result.Types.Contains("integer") && schemas[i].Types.Contains("integer") {
```
You also need to handle type `number` result.Types.Contains("number") alongwith integer

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,60 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+		} else if (result.Types.Contains("object") && schemas[i].Types.Contains("array")) || (result.Types.Contains("array") && schemas[i].Types.Contains("object")) {
+			return nil, fmt.Errorf("unable to merge these schemas")
+		} else if result.Types.Contains("string") && schemas[i].Types.Contains("string") {
+			return result, nil
```
Handling of `string`, `integer` and `boolean` isn't quite right. You are returning `result` as soon as you get `schemas[i]` that matches result type....what about the rest of the schemas??? 
Try this:

```
`{
	"$schema": "http://json-schema.org/draft-04/schema#",
	"type": "string",
	"title": "The b schema",
	"description": "An explanation about the purpose of this instance.",
	"allOf": [{
			"type": "string",
		},
		{
			"type": "string",
		},
		{
			"type": "boolean",
		}
	]
}`
```

This returns a merged type of `string` when it should have been failing with cannot merge because only schemas[0] is compared with schemas[1].....schemas[2], schemas[3] and so on are totally ignored

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,60 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+					result.ItemsChildren = append(result.ItemsChildren, schemas[i].ItemsChildren[j])
+				}
+			}
+		} else if (result.Types.Contains("object") && schemas[i].Types.Contains("array")) || (result.Types.Contains("array") && schemas[i].Types.Contains("object")) {
```
This if condition can be changed to 
```
if (result.Types.String() != schemas[i].Types.String()) {
  return nil, fmt.Errorf("unable to merge these schemas")
}
 ```
 
 because there are other types that can cause unmergeable status other than array and objects

aavarghese replied:
And this condition could be the first if condition check at line 911 in this series so that we dont go further with checking if types are different.

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,60 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
 }
 
+//Checks if a JSON schema contains a specific type
+func containsType(schema *gojsonschema.SubSchema, s string) bool {
```
This method doesn't look very useful if it is only used once and all it does is a String() comparison check

juliafriedman8 replied:
yeah it was something more complicated before that did not work haha and now you're right it is very simple--i removed it

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,60 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+				if result.ItemsChildren[j].Types.String() != schemas[i].ItemsChildren[j].Types.String() {
+					return nil, fmt.Errorf("unable to merge these schemas due to different types")
+				} else if !containsType(result.ItemsChildren[j], schemas[i].ItemsChildren[j].Types.String()) {
+					result.ItemsChildren = append(result.ItemsChildren, schemas[i].ItemsChildren[j])
```
It cannot be an append operation. Append add to the end of the array. Since order matters in an array tuple - the type has to be added at a specific location.

Also, it is possible that items children are of different lengths - so there is a fat chance of NPEs here when `j` element doesnt exist

aavarghese replied:
Try this good (should pass) example but it explodes with NPE: 

```
`{
	"type": "array",
	"allOf": [{
			"items": [{
					"type": "integer"
				},
				{
					"type": "integer"
				}
			]
		},
		{
			"items": [{
				"type": "integer"
			},
			{
				"type": "integer"
			},
			{
				"type": "string"
			}]
		}
	]
}`
```

aavarghese reviewed (commented):
Lots of comments. Sorry. Hopefully this helps.

ereslibre committed:
docs/kubernetes-primer: bump kubernetes resources versions (#3700)

* docs: bump `admission.k8s.io/v1beta1` to `admission.k8s.io/v1`
* docs: bump `Ingress` from `extensions` to `networking.k8s.io`

Bump the `Ingress` resources from the `extensions` API group to
`networking.k8s.io` and `v1` API version.

Signed-off-by: Rafael Fernández López <rfernandezlopez@suse.com>

ereslibre committed:
docs/kubernetes-primer: add mandatory `pathType` to Ingress resources (#3705)

Signed-off-by: Rafael Fernández López <rfernandezlopez@suse.com>

srenatus committed:
build: bump golang 1.16.5 -> 1.16.7 (#3710)

https://github.com/golang/go/issues?q=milestone%3AGo1.16.6+label%3ACherryPickApproved
https://github.com/golang/go/issues?q=milestone%3AGo1.16.7+label%3ACherryPickApproved

Signed-off-by: Stephan Renatus <stephan.renatus@gmail.com>

gujun4990 committed:
docs: Remove extra brackets in "Extending OPA" (#3712)

Signed-off-by: jun.gu <jun.gu@easystack.cn>

aavarghese commented on ast/compile.go:
```diff
@@ -928,6 +1032,30 @@ func parseSchema(schema interface{}) (types.Type, error) {
+			if (subSchema.Types.Contains("object") && allOfResult.Types.Contains("object")) || (subSchema.Types.Contains("array") && allOfResult.Types.Contains("array")) {
+				var temp []*gojsonschema.SubSchema
+				temp = append(temp, allOfResult)
+				temp = append(temp, subSchema)
```
You can combine the appends into one like this:
```
var temp []*gojsonschema.SubSchema
temp = append(temp, allOfResult, subSchema)
```

aavarghese commented on ast/compile.go:
```diff
@@ -928,6 +1032,30 @@ func parseSchema(schema interface{}) (types.Type, error) {
+				temp = append(temp, subSchema)
+				objectOrArrayResult, err := mergeSchemas(temp)
+				if err != nil {
+					return nil, fmt.Errorf("unable to merge these schemas due to: %w", err)
```
You can do a `return nil, err` instead of another fmt.Errorf

aavarghese commented on ast/compile.go:
```diff
@@ -928,6 +1032,30 @@ func parseSchema(schema interface{}) (types.Type, error) {
+				}
+				return parseSchema(objectOrArrayResult)
+			} else if subSchema.Types.String() != allOfResult.Types.String() {
+				return nil, fmt.Errorf("unable to merge these schemas due to: %w", err)
```
There is no `err` declared for this else if...so maybe do a ...
```
return nil, fmt.Errorf("unable to merge these schemas")
```

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,166 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+				schemaSet := NewSchemaSet()
+				schemaSet.Put(SchemaRootRef, schema)
+				c.WithSchemas(schemaSet)
+				compileStages(c, c.checkTypes)
```
You dont need to repeat all these lines of code again inside the `else`. To minimize code repetition, all you need is the `if tc.expectedError != nil {` right before the assert statements. Not the whole code...

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,110 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+				}
+			}
+		} else if result.Types.String() != schemas[i].Types.String() {
+			return nil, fmt.Errorf("unable to merge these schemas")
```
This condition is repeated twice...here and the first if check.

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,110 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+		if result.Types.Contains("object") && len(result.PropertiesChildren) > 0 && schemas[i].Types.Contains("object") && len(schemas[i].PropertiesChildren) > 0 {
+			for j := 0; j < len(schemas[i].PropertiesChildren); j++ {
+				result.PropertiesChildren = append(result.PropertiesChildren, schemas[i].PropertiesChildren[j])
+			}
```
You can shrink these 3 lines to one line:
```
result.PropertiesChildren = append(result.PropertiesChildren, schemas[i].PropertiesChildren...)
```
Notice the ... at the end that appends the entire slice

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,110 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+		if result.Types.String() != schemas[i].Types.String() {
+			return nil, fmt.Errorf("unable to merge these schemas")
+		}
+		if result.Types.Contains("object") && len(result.PropertiesChildren) > 0 && schemas[i].Types.Contains("object") && len(schemas[i].PropertiesChildren) > 0 {
```
Why is this not an else if?

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,110 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+			} else {
+				return nil, fmt.Errorf("unable to merge these schemas")
+			}
+		}
```
I dont think you need any of this code. Delete lines 925-981 and make change below too..
This is my reasoning:
If result has type string, any other type in schemas[i] will be rejected by the first if check and marked as unmergeable. If all are strings in schemas[i]. then result is still type string and thats all you need.


So you will need to remove the if else below on https://github.com/aavarghese/opa/pull/1/files?file-filters%5B%5D=.go#diff-067a842cc5e1a348551f4c7656120105acedafb2faac6e7b7ff57b75c5349570R983-R987
and just have a `return result, nil` --> you should return result for all result types not just object or array.

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,110 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+		return result, nil
+	} else {
+		return nil, fmt.Errorf("unable to merge these schemas")
+	}
```
Replace lines 983 to 987 with just a  `return result, nil`

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,110 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+	for i := 0; i < len(schemas); i++ {
+		if len(schemas[i].PropertiesChildren) > 0 {
+			if !schemas[i].Types.Contains("object") {
+				schemas[i].Types.Add("object")
```
See Go Lint test results here: https://github.com/aavarghese/opa/pull/1/checks?check_run_id=3272579440

Its asking you to make the following code style changes:

ast/compile.go:895:25: Error return value of `.Add` is not checked (errcheck)
				schemas[i].Types.Add("object")
				                    ^
ast/compile.go:899:25: Error return value of `.Add` is not checked (errcheck)
				schemas[i].Types.Add("array")

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,110 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+			}
+		} else if len(schemas[i].ItemsChildren) > 0 {
+			if !schemas[i].Types.Contains("array") {
+				schemas[i].Types.Add("array")
```
See Go Lint test results here: https://github.com/aavarghese/opa/pull/1/checks?check_run_id=3272579440

Its asking you to make the following code style changes:

ast/compile.go:895:25: Error return value of `.Add` is not checked (errcheck)
				schemas[i].Types.Add("object")
				                    ^
ast/compile.go:899:25: Error return value of `.Add` is not checked (errcheck)
				schemas[i].Types.Add("array")

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,110 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+					result.ItemsChildren = append(result.ItemsChildren, schemas[i].ItemsChildren[j])
+				}
+				if result.ItemsChildren[j].Types.String() != schemas[i].ItemsChildren[j].Types.String() {
+					return nil, fmt.Errorf("unable to merge these schemas due to different types")
```
Can you remove the "due to different types" to keep it consistent with all the other err messages youhave?

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,110 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+				if result.ItemsChildren[j].Types.String() != schemas[i].ItemsChildren[j].Types.String() {
+					return nil, fmt.Errorf("unable to merge these schemas due to different types")
+				} else if result.ItemsChildren[j].Types.String() == schemas[i].ItemsChildren[j].Types.String() {
+					result.ItemsChildren[j].Types = schemas[i].ItemsChildren[j].Types
```
Do we need this else if and setting .Types?
result already has the same matching type as schemas[i], so why set it again?

srenatus committed:
docs/wasm: clarify opa_eval args (#3699)

Signed-off-by: Stephan Renatus <stephan.renatus@gmail.com>

srenatus committed:
rego: pass inter-query builtin cache to wasm sdk (#3721)

Found when looking into #3716.

With this change, multiple calls to a wasm-backed eval that uses http.send
would re-use the results (if caching permits).

Signed-off-by: Stephan Renatus <stephan.renatus@gmail.com>

aavarghese commented on ast/compile_test.go:
```diff
@@ -875,6 +875,139 @@ func TestCompilerCheckTypesWithSchema(t *testing.T) {
+			compileStages(c, c.checkTypes)
+			if tc.expectedError != nil {
+				errors := c.Errors
+				if !strings.Contains(errors.Error(), tc.expectedError.Error()) {
```
You could change this to `if !strings.Contains(c.Errors.Error(), tc.expectedError.Error()) {` and remove the previous line `errors := c.Errors`

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,47 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+			if !schemas[i].Types.Contains("object") {
+				if err := schemas[i].Types.Add("object"); err == nil {
+					schemas[i].Types.Add("object")
+				}
```
Actually, the check needs to for `err != nil` <-- thats the concerning case.
The first Add (inside the if) already adds the the object type so we dont need to do it twice.

So like this:
```
if err := schemas[i].Types.Add("object"); err != nil {
	return nil, fmt.Errorf("unable to set the type in schemas")
}
```

aavarghese commented on ast/compile.go:
```diff
@@ -883,6 +883,47 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+			if !schemas[i].Types.Contains("array") {
+				if err := schemas[i].Types.Add("array"); err == nil {
+					schemas[i].Types.Add("array")
+				}
```
Same as above

aavarghese commented on docs/content/schemas.md:
````diff
@@ -529,6 +529,81 @@ Once this is fixed, the second typo is highlighted, prompting the user to choose
+	              ^
+	              have: "servers"
+	              want: ["server"]
+```
````
This needs to be fixed. This isn't the output exactly from `opa eval`

aavarghese commented on docs/content/schemas.md:
````diff
@@ -529,6 +529,81 @@ Once this is fixed, the second typo is highlighted, prompting the user to choose
+	                     have: "versions"
+	                     want: ["version"]
+```
+Because the properties `kind`, `version`, and `accessNum` are all under the `allOf` keyword, the resulting schema that the given data must be validated against will contain the types contained in these properties children (string and integer). 
````
Same as above.

srenatus committed:
wasm: update wasmtime-go, inspect trap code (#3704)

* deps: bump wasmtime-go (0.28.0 -> 0.29.0)

This gets us wasmtime 0.29.0,

https://github.com/bytecodealliance/wasmtime/blob/v0.29.0/RELEASES.md#0290

and these changes to wasmtime-go:

https://github.com/bytecodealliance/wasmtime-go/compare/v0.28.0...v0.29.0

* wasm/vm: check trap code when constructing error message

With this, non-interrupt traps will not look like interrupts.

Signed-off-by: Stephan Renatus <stephan.renatus@gmail.com>

koponen-styra committed:
topdown: Allow store return ast.Values. (#3703)

This allows a custom store implementation to return ast.Values as
such.

Signed-off-by: Teemu Koponen <koponen@styra.com>

Tej-Singh-Rana committed:
docs/content/policy-language.md: changed 'As as result' to 'As a result' (#3727)

Signed-off-by: Tej Singh Rana <iamtejsinghrana@gmail.com>

srenatus committed:
Merge branch 'main' into allOf-2
